fvm/evm/emulator/state/base.go

```diff
@@ -0,0 +1,472 @@
+	if err != nil {
+		return err
+	}
+
+	// 4. remove the code
+	if acc.HasCode() {
+		err = v.storeCode(addr, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	// 5. remove storage slots
+	if len(acc.CollectionID) > 0 {
+		col, found := v.slots[addr]
+		if !found {
+			col, err = v.collectionProvider.CollectionByID(acc.CollectionID)
+			if err != nil {
+				return err
+			}
+		}
+		// delete all slots related to this account (eip-6780)
+		err = col.Destroy()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Commit commits the changes to the underlying storage layers
+func (v *BaseView) Commit() error {
+	// commit collection changes
+	err := v.collectionProvider.Commit()
+	if err != nil {
+		return err
+	}
+
+	// if this is the first time we are setting up an
+	// account collection, store its collection id.
+	if v.accountSetupOnCommit {
+		err = v.ledger.SetValue(v.rootAddress[:], []byte(AccountsStorageIDKey), v.accounts.CollectionID())
+		if err != nil {
+			return err
+		}
+		v.accountSetupOnCommit = false
+
+	}
+
+	// if this is the first time we are setting up an
+	// code collection, store its collection id.
+	if v.codeSetupOnCommit {
+		err = v.ledger.SetValue(v.rootAddress[:], []byte(CodesStorageIDKey), v.codes.CollectionID())
+		if err != nil {
+			return err
+		}
+		v.codeSetupOnCommit = false
+	}
+	return nil
+}
+
+func (v *BaseView) fetchOrCreateCollection(path string) (collection *Collection, created bool, error error) {
+	collectionID, err := v.ledger.GetValue(v.rootAddress[:], []byte(path))
+	if err != nil {
+		return nil, false, err
+	}
+	if len(collectionID) == 0 {
+		collection, err = v.collectionProvider.NewCollection()
+		return collection, true, err
+	}
+	collection, err = v.collectionProvider.CollectionByID(collectionID)
+	return collection, false, err
+}
+
+func (v *BaseView) getAccount(addr gethCommon.Address) (*Account, error) {
+	// check cached accounts first
+	acc, found := v.cachedAccounts[addr]
+	if found {
+		return acc, nil
+	}
+
+	// then collect it from the account collection
+	data, err := v.accounts.Get(addr.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	// decode it
+	acc, err = DecodeAccount(data)
+	if err != nil {
+		return nil, err
+	}
+	// cache it
+	if acc != nil {
+		v.cachedAccounts[addr] = acc
+	}
+	return acc, nil
+}
+
+func (v *BaseView) storeAccount(acc *Account) error {
+	data, err := acc.Encode()
+	if err != nil {
+		return err
+	}
+	// update the cache
+	v.cachedAccounts[acc.Address] = acc
+	return v.accounts.Set(acc.Address.Bytes(), data)
+}
+
+func (v *BaseView) getCode(addr gethCommon.Address) ([]byte, error) {
+	// check the cache first
+	code, found := v.cachedCodes[addr]
+	if found {
+		return code, nil
+	}
+	// check if account exist and has codeHash
+	acc, err := v.getAccount(addr)
+	if err != nil {
+		return nil, err
+	}
+	// if no account found return
+	if acc == nil {
+		return nil, nil
+	}
+	// check account has code
+	if !acc.HasCode() {
+		return nil, nil
+	}
```

Comment:
isn't it the getAccount does that already?

Reply:
> isn't it the getAccount does that already?

Yes, `getAccount` also loads account from `accounts` collection if it isn't cached.

Since we only want to get `code` in this function and code is stored in `codes` collection, it probably isn't necessary to load account when it isn't cached.

Reply:
right right.